docs/reference/database-migration.md

```diff
@@ -27,4 +27,31 @@ Under the hood, our KV implementation relies on a backing DB, which persists the
+
+### Optimistic Locking with KV
+
+One important key difference between SQL databases and Key Value Store is the ability to lock resources. While this is a common practice with relational databases, Key Value stores lack this ability. Let us take a look at a common lakeFS flow, a `Commit`. During this flow, several ref store operations are performed:
+* All relevant (`Branch` correlated) uncommitted objects are collected and marked as committed 
+* A new `Commit` object is created
+* The relevant `Branch` is updated to point to the new commit
+The `Commit` flow includes multiple database accesses and modifications, and is very sensitive to concurrent executions: If 2 `Commit` flows run in parallel, we cannot control which one will update the `Branch` last and we cannot guarantee correctness of the data. `lakeFS` with PostgreSQL solved it simply by locking the `Branch` for the entire `Commit` operation, which prevented concurrent execution of such flows.
+Now, with KV Store replacing the SQL database, this easy solution is no longer available. Instead, we implemented an [Optimistic Locking](https://en.wikipedia.org/wiki/Optimistic_concurrency_control) algorithm, which leverages the KV Store `Compare-and-Set` (`CaS`) functionality to remember the `Branch` state at the beginning of the `Commit` flow, and updating the branch at the end, only if it remains unchanged, using `CaS`. Before sampling the `Branch` another update, of the `Branch`'s `StagingToken`, is made, and so, if another `Commit` is already running in parallel, it will fail the final step, of updating the `Branch`, as it was already changed. If 2 commits do run in parallel there are only 2 possible outcomes:
+1. Either 1 of the `Commits` manages to start, sample the `Branch` state, modify the `StagingToken`, create a new `Commit` and use `Cas` to find out that the branch was not modified and update it, and so it succeeds and causing the other `Commit` to fail, or
+2. Both `Commits` interfere with each other, and both will fail, as  depicted in the following scenario:
+  * `Commit` A sets the `StagingToken` to val1 and samples the `Branch`,
+  * `Commit` B sets the `StagingToken` to val2 and samples the `Branch`,
+  * `Commit` A finishes, tries to update the `Branch` and fails due to the recent modification by `Commit` B,
+  * `Commit` A is retired, setting `StagingToken` to val3 and samples the `Branch`
+  * `Commit` B finishes, tries to update the `Branch` and fails due to the recent modification by `Commit` A,
+  * And so on...
+At some point, as retries will be exhausted, commits will succeed, if not earlier, but this was definitely something that had to be taken into consideration when designing our KV Store. You can read more on the Commit Flow in the [dedicated section in the KV Design](https://github.com/treeverse/lakeFS/blob/master/design/accepted/metadata_kv/index.md#graveler-metadata---branches-and-staged-writes)
```

Comment:
When phrased like this it is very worrying, and makes it sound as though workflows that attempt to commit periodically will never work.
At the very least I would explain that commits are idempotent, so as soon as one succeeds we advance, and other commits only need to commit work started after that successful commit.
Otherwise starvation would be inevitable.

Reply:
@arielshaqed - AFAICT, starvation is possible (probably not inevitable).
The best thing I could come up with is the remove the last bullet "and so on..." which implies endlessness, and change the next line to:
"The first commit to succeed, releases this mutual interference, and the next commits can continue uninterrupted."

WDYT?
I wuold love to adopt any better suggestion

Reply:
Either leave it alone ("eventually a commit succeeds"), or point to #3586.  The current makes me feel really bad -- it gives little hope, but also doesn't say we know how to fix it when that becomes necessary.  So I'd prefer to go in either direction.

Reply:
@arielshaqed - rephrased as we eariler discessed